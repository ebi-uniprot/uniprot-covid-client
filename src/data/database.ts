import EntrySection from '../model/types/EntrySection';
import { DatabaseCategory, DatabaseInfo } from '../model/types/DatabaseRefs';
import {
  getDatabaseInfoMaps,
  selectDatabases,
  getEntrySectionToDatabaseCategoryOrder,
} from '../utils/database';
import databaseInfoJson from './databaseInfo.json';
import externalUrls from '../utils/externalUrls';

const databaseInfo: DatabaseInfo = databaseInfoJson;

export const databaseCategoryToString = {
  [DatabaseCategory.CHEMISTRY]: 'Chemistry',
  [DatabaseCategory.DOMAIN]: 'Family and domain databases',
  [DatabaseCategory.EXPRESSION]: 'Gene expression databases',
  [DatabaseCategory.FAMILY]: 'Protein family/group databases',
  [DatabaseCategory.GEL]: '2D gel databases',
  [DatabaseCategory.GENOME]: 'Genome annotation databases',
  [DatabaseCategory.INTERACTION]: 'Protein-protein interaction databases',
  [DatabaseCategory.ORGANISM]: 'Organism-specific databases',
  [DatabaseCategory.OTHER]: 'Other',
  [DatabaseCategory.PATHWAY]: 'Enzyme and pathway databases',
  [DatabaseCategory.PHYLOGENOMIC]: 'Phylogenomic databases',
  [DatabaseCategory.POLYMORPHISM]: 'Polymorphism and mutation databases',
  [DatabaseCategory.PROTEOMIC]: 'Proteomic databases',
  [DatabaseCategory.PROTOCOL]: 'Protocols and materials databases',
  [DatabaseCategory.PTM]: 'PTM databases',
  [DatabaseCategory.SEQUENCE]: 'Sequence databases',
  [DatabaseCategory.STRUCTURE]: '3D structure databases',
};

export const {
  databaseCategoryToNames,
  databaseNameToCategory,
  databaseToDatabaseInfo,
  implicitDatabaseXRefs,
} = getDatabaseInfoMaps(databaseInfo);

export const PDBMirrors = ['PDB', 'RCSB-PDB', 'PDBj', 'PDBsum'];

export const PDBMirrorsInfo = PDBMirrors.map(
  PDBMirror => databaseToDatabaseInfo[PDBMirror]
);

const databaseSelector = selectDatabases(databaseCategoryToNames);

export const entrySectionToDatabaseNames = new Map<EntrySection, string[]>();
entrySectionToDatabaseNames.set(
  EntrySection.Expression,
  databaseSelector({
    categories: [DatabaseCategory.EXPRESSION],
    whitelist: ['HPA'],
  })
);
entrySectionToDatabaseNames.set(
  EntrySection.FamilyAndDomains,
  databaseSelector({
    categories: [DatabaseCategory.PHYLOGENOMIC, DatabaseCategory.DOMAIN],
    whitelist: [
      'MobiDB', // Implicit
      'ProtoNet', // Implicit
      'GPCRDB', // Implicit
    ],
  })
);
entrySectionToDatabaseNames.set(
  EntrySection.Function,
  databaseSelector({
    categories: [DatabaseCategory.PATHWAY, DatabaseCategory.FAMILY],
    whitelist: ['SwissLipids'],
  })
);
entrySectionToDatabaseNames.set(
  EntrySection.Interaction,
  databaseSelector({
    categories: [DatabaseCategory.INTERACTION],
    whitelist: ['BindingDB'],
  })
);
entrySectionToDatabaseNames.set(EntrySection.NamesAndTaxonomy, [
  'ArachnoServer',
  'Araport',
  'CGD',
  'ConoServer',
  'dictyBase',
  'EcoGene',
  'EuPathDB',
  'FlyBase',
  'Gramene',
  'HGNC',
  'LegioList',
  'Leproma',
  'MaizeGDB',
  'MGI',
  'MIM',
  'neXtProt',
  'PomBase',
  'PseudoCAP',
  'RGD',
  'SGD',
  'TAIR',
  'TubercuList',
  'VGNC',
  'WormBase',
  'Xenbase',
  'ZFIN',
]);
entrySectionToDatabaseNames.set(EntrySection.PathologyAndBioTech, [
  'DisGeNET',
  'GeneReviews',
  'MalaCards',
  'MIM',
  'OpenTargets',
  'Orphanet',
  'PharmGKB',
  'ChEMBL',
  'DrugBank',
  'GuidetoPHARMACOLOGY',
  'BioMuta',
  'DMDM',
  'Allergome',
]);
entrySectionToDatabaseNames.set(
  EntrySection.ProteinProcessing,
  databaseSelector({
    categories: [
      DatabaseCategory.PROTEOMIC,
      DatabaseCategory.GEL,
      DatabaseCategory.PTM,
    ],
    whitelist: ['PMAP-CutDB'],
  })
);
entrySectionToDatabaseNames.set(
  EntrySection.Sequence,
  databaseSelector({
    categories: [DatabaseCategory.SEQUENCE, DatabaseCategory.GENOME],
  })
);
entrySectionToDatabaseNames.set(
  EntrySection.Structure,
  databaseSelector({
    categories: [DatabaseCategory.STRUCTURE],
    whitelist: [
      'EvolutionaryTrace',
      'ModBase', // Implicit
    ],
  })
);

// This is used to catch those that aren't listed in the page sections
entrySectionToDatabaseNames.set(
  EntrySection.ExternalLinks,
  databaseSelector({
    categories: [DatabaseCategory.OTHER, DatabaseCategory.PROTOCOL],
    whitelist: [
      'HUGE', // Implicit
      'ROUGE', // Implicit
      'GenAtlas', // Implicit
      ...PDBMirrors,
    ],
  })
);

export const getDatabaseNameToEntrySection = (
  databaseName: string
): EntrySection | undefined => {
  let entrySection;
  entrySectionToDatabaseNames.forEach((value, key) => {
    if (value.includes(databaseName)) {
      entrySection = key;
    }
  });
  return entrySection;
};

export const entrySectionToDatabaseCategoryOrder = getEntrySectionToDatabaseCategoryOrder(
  entrySectionToDatabaseNames,
  databaseNameToCategory
);

export const getDatabaseInfoByName = (dbName: string) =>
  databaseInfo.find(
    dbInfo => dbInfo.name.toLowerCase() === dbName.toLowerCase()
  );

// If each of the keys are present then show the values
export const implicitDatabaseDRPresence: { [key: string]: string[] } = {
  // these EMBL mirrors are taken care of in xrefview as they are displayed differently
  // EMBL: ['GenBank', 'DDBJ'],
  PDB: ['PDBe-KB'], // eg P05067
  MIM: ['SOURCE_MIM'], // eg P05067
  MGI: ['SOURCE_MGI'], // eg E9PXF8
  HGNC: ['GenAtlas'], // eg Q9Y263
};

// If each of the keys are not present then show the value
export const implicitDatabaseDRAbsence: { [key: string]: string[] } = {
  SMR: ['SWISS-MODEL-Workspace'], // eg P16646
};

export const implicitDatabaseAlwaysInclude = [
  'ModBase', // eg P05067
  'MobiDB', // eg P05067
  'ProtoNet', // eg P05067
];

export const implicitDatabaseGenePatternOrganism = {
  pattern: /KIAA\d{4}/i,
  organism: {
    Human: 'HUGE', // eg Q96PV4
    Mouse: 'ROUGE', // eg Q8CJ19
  },
};

export const implicitDatabaseSimilarityComment = {
  GPCRDB: 'Belongs to the G-protein coupled receptor', // eg Q7RTX1
};

export const implicitDatabasesEC = ['ENZYME']; // eg Q54WR4

export const viewProteinLinkDatabases = new Map([
  ['InterPro', externalUrls.InterPro],
  ['Pfam', externalUrls.Pfam],
  ['SMART', externalUrls.SMART],
  ['PROSITE', externalUrls.PROSITE],
]);
