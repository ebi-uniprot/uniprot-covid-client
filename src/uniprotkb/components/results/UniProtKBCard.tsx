/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { FC, Fragment } from 'react';
import idx from 'idx';
import { Card } from 'franklin-sites';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { UniProtkbAPIModel } from '../../adapters/uniProtkbConverter';
import { getKeywordsForCategories } from '../../utils/KeywordsUtil';
import KeywordCategory from '../../types/keywordCategory';
import { KeywordList } from '../protein-data-views/KeywordView';
import UniProtKBTitle from '../protein-data-views/UniProtKBTitle';
import AnnotationScoreDoughnutChart, {
  DoughnutChartSize,
} from '../protein-data-views/AnnotationScoreDoughnutChart';
import getProteinHighlights from '../../adapters/proteinHighlights';
import './styles/uniprotkb-card.scss';

const UniProtKBCard: FC<
  {
    data: UniProtkbAPIModel;
    selected: boolean;
    handleEntrySelection: (rowId: string) => void;
  } & RouteComponentProps
> = ({ data, selected, handleEntrySelection, history }): JSX.Element => {
  let recommendedNameNode;
  const recommendedName = idx(
    data,
    (_): string => _.proteinDescription.recommendedName.fullName.value
  );
  if (recommendedName) {
    recommendedNameNode = `${recommendedName} · `;
  }

  const highlights = getProteinHighlights(data);

  const organismNameNode = (
    <Fragment>
      {idx(data, (_): string => _.organism.scientificName)}
      {' · '}
    </Fragment>
  );

  let geneNameListNode;
  if (data.genes) {
    geneNameListNode = (
      <Fragment>
        {'Gene: '}
        {data.genes
          .filter((geneName) => geneName.geneName)
          .map((geneName) => geneName.geneName && geneName.geneName.value)
          .join(', ')}
        {' · '}
      </Fragment>
    );
  }

  const sequenceLengthNode = `${data.sequence.length} amino-acids · `;

  const { annotationScore } = data;
  const annotationScoreNode = (
    <AnnotationScoreDoughnutChart
      score={annotationScore}
      size={DoughnutChartSize.small}
    />
  );

  let keywordsNode;
  if (data.keywords) {
    const categorisedKewywords = getKeywordsForCategories(data.keywords, [
      KeywordCategory.MOLECULAR_FUNCTION,
      KeywordCategory.BIOLOGICAL_PROCESS,
      KeywordCategory.DISEASE,
    ]);

    if (categorisedKewywords.length > 0) {
      keywordsNode = categorisedKewywords.map((keywordCategory, index) => (
        <Fragment key={keywordCategory.category}>
          {index > 0 && ' · '}
          <KeywordList keywords={keywordCategory.keywords} />
        </Fragment>
      ));
    }
  }

  return (
    <Card
      links={highlights}
      onClick={() => history.push(`/uniprotkb/${data.primaryAccession}`)}
    >
      <section className="uniprot-card">
        <section className="uniprot-card__left">
          <input
            type="checkbox"
            checked={selected}
            onClick={(e) => e.stopPropagation()}
            onChange={() => handleEntrySelection(data.primaryAccession)}
            data-testid="up-card-checkbox"
          />
        </section>
        <section className="uniprot-card__right">
          <h5>
            <UniProtKBTitle
              primaryAccession={data.primaryAccession}
              entryType={data.entryType}
              uniProtkbId={data.uniProtkbId}
            />
          </h5>
          <section>
            {recommendedNameNode}
            {organismNameNode}
            {geneNameListNode}
            {sequenceLengthNode}
            {annotationScoreNode}
          </section>
          <section>
            <small>{keywordsNode}</small>
          </section>
        </section>
      </section>
    </Card>
  );
};

export default withRouter(UniProtKBCard);
