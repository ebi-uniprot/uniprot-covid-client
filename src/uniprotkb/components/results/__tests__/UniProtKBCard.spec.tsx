import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';
import UniProtKBCard from '../UniProtKBCard';
import data from '../../../__mocks__/entryModelData.json';

const handleEntrySelection = jest.fn();

let item;

describe('UniProtKBCard component', () => {
  beforeEach(() => {
    item = render(
      <Router>
        <UniProtKBCard
          data={data}
          selectedEntries={{}}
          handleEntrySelection={handleEntrySelection}
        />
      </Router>
    );
  });

  test('should render', () => {
    const { asFragment } = item;
    expect(asFragment()).toMatchSnapshot();
  });

  test('should select a row', () => {
    const { getByTestId } = item;
    const checkbox = getByTestId('up-card-checkbox');
    fireEvent.click(checkbox);
    expect(handleEntrySelection).toHaveBeenCalled();
  });
});
