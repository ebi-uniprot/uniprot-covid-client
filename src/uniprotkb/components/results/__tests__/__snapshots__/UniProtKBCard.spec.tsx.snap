// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniProtKBCard component should render 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <section
      class=" card--has-hover"
      role="button"
      tabindex="0"
    >
      <div
        class="card__content"
      >
        <section
          class="uniprot-card"
        >
          <section
            class="uniprot-card__left"
          >
            <input
              data-testid="up-card-checkbox"
              type="checkbox"
            />
          </section>
          <section
            class="uniprot-card__right"
          >
            <h5>
              <span
                class="uniprot-title"
              >
                <span
                  class="uniprot-title__status icon--unreviewed"
                >
                  <svg
                    fill="currentColor"
                    viewBox="0 0 400 512"
                  >
                    <path
                      d="M400 144H256V0l144 144zm-176 32V0H0v512h400V176H224z"
                    />
                  </svg>
                </span>
                P21802 · uniprot id
              </span>
            </h5>
            <section>
              rec full Name · scientific name · Gene: some Gene · 10 amino-acids · 
              <span
                title="Annotation Score"
              >
                <span
                  class="doughnut-chart--small colour-platinum"
                >
                  <span
                    class="doughnut-chart--small__left-wrap"
                  >
                    <span
                      class="doughnut-chart--small__left-wrap__loader colour-sea-blue"
                      style="transform: rotate(0deg);"
                    />
                  </span>
                  <span
                    class="doughnut-chart--small__right-wrap"
                  >
                    <span
                      class="doughnut-chart--small__right-wrap__loader colour-sea-blue"
                      style="transform: rotate(72deg);"
                    />
                  </span>
                  <span
                    class="doughnut-chart--small__inner-circle"
                  >
                    1/5
                  </span>
                </span>
              </span>
            </section>
            <section>
              <small />
            </section>
          </section>
        </section>
      </div>
    </section>
    <section
      class="card__actions"
    >
      <span
        class="card__link false"
      >
        <a
          href="/uniprotkb/P21802#Sequence"
        >
          1 isoform
        </a>
      </span>
      <span
        class="card__link card__link--separator"
      >
        <a
          href="/uniprotkb/P21802#Interaction"
        >
          2 interactions
        </a>
      </span>
      <span
        class="card__link card__link--separator"
      >
        <a
          href="/uniprotkb/P21802#Pathology & BioTech"
        >
          1 disease
        </a>
      </span>
      <span
        class="card__link card__link--separator"
      >
        <a
          href="/uniprotkb/P21802/publications"
        >
          7 publications
        </a>
      </span>
    </section>
  </div>
</DocumentFragment>
`;
